### PR TITLE
Fix race conditions when running multiple cells in native interactive window from codelens

### PR DESCRIPTION
**Pull request by joyceerhl**
For https://github.com/microsoft/vscode-jupyter/issues/6494 and partially https://github.com/microsoft/vscode-jupyter/issues/6603

### CODE REVIEW

File: src/client/datascience/interactive-window/nativeInteractiveWindowProvider.ts [hunk 1]
```diff
@@ -109,69 +110,83 @@ export class NativeInteractiveWindowProvider implements IInteractiveWindowProvid
...
+        if (resource === undefined) {
+            return this.createInteractiveWindowPromise(resource, mode);
+        }
 
-        return result;
+        const existingPromise = this.mapOfResourcesToInteractiveWindowPromises.get(resource.toString());
+        // If there's a resource and no existing promise, create a new one
+        if (existingPromise === undefined) {
+            const promise = this.createInteractiveWindowPromise(resource, mode);
+            this.mapOfResourcesToInteractiveWindowPromises.set(resource.toString(), promise);
```
Comment (joyceerhl) [inline src/client/datascience/interactive-window/nativeInteractiveWindowProvider.ts:184]: Previously interactive window creation was asynchronous and could take a while, so if you ran a bunch of cells it was possible for the previous interactive window to still be in the process of initializing, and this caused us to try and create two interactive windows for the same submitting file, which uncovered a race condition in VS Code where the editor input might not have a notebook model initialized yet. This ensures that we reuse the existing init promise if a bunch of requests from the same file come in at once.

File: src/client/datascience/interactive-window/nativeInteractiveWindow.ts [hunk 2]
```diff
@@ -664,14 +664,13 @@ export class NativeInteractiveWindow implements IInteractiveWindowLoadable {
...
         };
-        edit.replaceNotebookCells(
-            this.notebookDocument.uri,
-            new NotebookRange(this.notebookDocument.cellCount, this.notebookDocument.cellCount),
-            [
-                notebookCellData // TODO generalize to arbitrary languages and cell types
-            ]
-        );
-        await workspace.applyEdit(edit);
+        await chainWithPendingUpdates(this.notebookDocument, (edit) => {
```
Comment (joyceerhl) [inline src/client/datascience/interactive-window/nativeInteractiveWindow.ts:667]: Cells were getting inserted into the `NotebookDocument` out of order, needed `chainWithPendingUpdates` to enforce edits in the right sequence

**Comment by joyceerhl** [comment]
There's still a problem which also exists for the webview interactive window that I don't fully understand where, even though `addCode` is getting invoked with the right sequence of code cells, `executeObservable` seems to receive them out of order, i.e. if I do `print(1) ... print(4)`, the actual Jupyter session is told to `print(2); print(3); print(4); print(1)`, so the execution counts are all wrong because the execution requests were received in the wrong order. Might need your help with this @rchiodo

File: src/client/datascience/interactive-window/nativeInteractiveWindowProvider.ts [hunk 3]
```diff
@@ -109,69 +110,83 @@ export class NativeInteractiveWindowProvider implements IInteractiveWindowProvid
...
+                return result;
+            });
+    }
 
-        // fire created event
-        this._onDidCreateInteractiveWindow.fire(result);
+    protected async create(resource: Resource, mode: InteractiveWindowMode): Promise<NativeInteractiveWindow> {
+        // If there's no resource, just create a new one
+        if (resource === undefined) {
+            return this.createInteractiveWindowPromise(resource, mode);
```
Comment (rchiodo) [inline src/client/datascience/interactive-window/nativeInteractiveWindowProvider.ts:177]: Could this cause the same race? You might create two interactive windows?

Wondering if the undefined case should also have an entry in the map

**Review by rchiodo — APPROVED** [review]

**Comment by rchiodo** [comment]
> There's still a problem which also exists for the webview interactive window that I don't fully understand where, even though `addCode` is getting invoked with the right sequence of code cells, `executeObservable` seems to receive them out of order, i.e. if I do `print(1) ... print(4)`, the actual Jupyter session is told to `print(2); print(3); print(4); print(1)`, so the execution counts are all wrong because the execution requests were received in the wrong order. Might need your help with this @rchiodo

My guess is that the async code in addCode is causing one of them to wait but not another.

File: src/client/datascience/interactive-window/nativeInteractiveWindowProvider.ts (hunk 3 above)
Comment (joyceerhl) [reply src/client/datascience/interactive-window/nativeInteractiveWindowProvider.ts:177]: If resource is undefined that means there's no backing script file, i.e. being created from command palette, in which case we would always want to create a new interactive window, no?

Comment (rchiodo) [reply src/client/datascience/interactive-window/nativeInteractiveWindowProvider.ts:177]: Looking at the old code, it was async too, but the creation of the window was based entirely on the mode. Why does the controller need to be created before the window?

Comment (rchiodo) [reply src/client/datascience/interactive-window/nativeInteractiveWindowProvider.ts:177]: The old code would just await the async parts after creation. Maybe we can live chat instead?

**Comment by joyceerhl** [comment]
@rchiodo I'll fix up the promise chaining issue we talked about which impacts webview and explore refactoring the ctor in a separate PR

**Pull request merged**